prober/derp.go

```diff
@@ -577,6 +602,246 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+	if err != nil {
+		return fmt.Errorf("failed to configure tun: %w", err)
+	}
+
+	// Depending on platform, we need some space for headers at the beginning
+	// of TUN packets. The below constant is more than enough space for any
+	// platform that this might run on.
+	tunStartOffset := device.MessageTransportHeaderSize
+
+	// Read packets from TUN device and forward all IPv4 packets destined to
+	// our listener's address via DERP.
+	tunReadC := make(chan error, 1)
+	go func() {
+		numBufs := wgconn.IdealBatchSize
+		bufs := make([][]byte, 0, numBufs)
+		sizes := make([]int, numBufs)
+		for range numBufs {
+			bufs = append(bufs, make([]byte, mtu+tunStartOffset))
+		}
+
+		src, dst := make([]byte, 4), make([]byte, 4)
+		for {
+			n, err := dev.Read(bufs, sizes, tunStartOffset)
+			if err != nil {
+				return
+			}
+
+			for i := range n {
+				pkt := bufs[i][tunStartOffset : sizes[i]+tunStartOffset]
+				// Skip everything except valid IPv4 packets
+				if len(pkt) < 20 {
+					// Doesn't even have a full IPv4 header
+					continue
+				}
+				if pkt[0]>>4 != 4 {
+					// Not IPv4
+					continue
+				}
+
+				destinationAddrBytes := destinationAddr.As4()
+				if !bytes.Equal(pkt[16:20], destinationAddrBytes[:]) {
+					// Unexpected dst address
+					continue
+				}
+
+				copy(src, pkt[12:16])
+				copy(dst, pkt[16:20])
+				copy(pkt[12:16], dst)
+				copy(pkt[16:20], src)
+
+				if err := fromc.Send(toc.SelfPublicKey(), pkt); err != nil {
+					tunReadC <- err
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvC := make(chan error, 1)
+	go func() {
+		buf := make([]byte, mtu+tunStartOffset)
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvC <- fmt.Errorf("failed to receive: %w", err)
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				if v.Source != fromc.SelfPublicKey() {
+					recvC <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				pkt := v.Data
+				copy(buf[tunStartOffset:], pkt)
+				if _, err := dev.Write([][]byte{buf[:len(pkt)+tunStartOffset]}, tunStartOffset); err != nil {
+					recvC <- fmt.Errorf("failed to write to TUN device: %w", err)
+					return
+				}
+			case derp.KeepAliveMessage:
+				// Silently ignore.
+			default:
+				log.Printf("%v: ignoring Recv frame type %T", to.Name, v)
+				// Loop.
+			}
+		}
+	}()
+
+	// Start a listener to receive the data
+	l, err := net.Listen("tcp", fmt.Sprintf("%s:0", ifAddr))
+	if err != nil {
+		return fmt.Errorf("failed to listen: %s", err)
+	}
+	defer l.Close()
+
+	randData := make([]byte, 128000)
+	_, err = crand.Read(randData)
+	if err != nil {
+		return fmt.Errorf("failed to initialize random data: %w", err)
+	}
+
+	// Read the data we've sent
+	readC := make(chan error, 1)
+	go func() {
+		readConn, err := l.Accept()
+		if err != nil {
+			readC <- err
+		}
+		defer readConn.Close()
+		_, err = io.CopyN(io.Discard, readConn, size)
+		readC <- err
```

Comment:
This is where we'd want to set `transferTime` or similar, when the read is complete. If accessing `transferTime` in this goroutine we need to ensure we don't create a data race with any other accesses.